autopush/web/registration.py

```diff
@@ -34,242 +32,356 @@
+        """Write the JSON response of the created endpoint"""
+        response = dict(channelID=chid, endpoint=endpoint)
+        if new_uaid:
+            secret = None
+            if self.ap_settings.bear_hash_key:
+                secret = generate_hash(
+                    self.ap_settings.bear_hash_key[0], uaid.hex)
+            response.update(uaid=uaid.hex, secret=secret)
+            # Apply any router specific fixes to the outbound response.
+            router = self.ap_settings.routers[router_type]
+            router.amend_endpoint_response(response, router_data)
+        self.write(json.dumps(response))
+        self.log.debug("Endpoint registered via HTTP",
+                       client_info=self._client_info)
+        self.finish()
+
+    def _success(self, result):
+        """Writes out empty 200 response"""
+        self.write({})
+        self.finish()
+
+
+class NewRegistrationHandler(BaseRegistrationHandler):
+    """Handle new bridge uaid registrations"""
+    cors_methods = "POST"
+
+    @threaded_validate(NewRegistrationSchema)
+    def post(self, router_type, router_data):
+        # type: (str, JSONDict) -> Deferred
         """HTTP POST
 
-        Endpoint generation and optionally router type/data registration.
+        Router type/data registration.
 
         """
         self.add_header("Content-Type", "application/json")
         self.ap_settings.metrics.increment("updates.client.register",
                                            tags=self.base_tags())
 
-        # If the client didn't provide a CHID, make one up.
-        # Note, RegistrationSchema may explicitly set "chid" to None
-        # THIS VALUE MUST MATCH WHAT'S SPECIFIED IN THE BRIDGE CONNECTIONS.
-        # currently hex formatted.
-        if not chid:
-            chid = uuid.uuid4().hex
-        rinfo.data["channelID"] = chid
-
-        if not uaid:
-            uaid = uuid.uuid4()
-            d = defer.execute(rinfo.register, uaid, uri=self.request.uri)
-            d.addCallback(
-                lambda _:
-                deferToThread(self._register_user_and_channel,
-                              uaid, chid, rinfo)
-            )
-            d.addCallback(
-                self._write_endpoint, uaid, chid, rinfo, new_uaid=True)
-            d.addErrback(self._router_fail_err)
-            d.addErrback(self._response_err)
-        else:
-            d = deferToThread(self._register_channel,
-                              uaid, chid, rinfo.app_server_key)
-            d.addCallback(self._write_endpoint, uaid, chid, rinfo)
-            d.addErrback(self._response_err)
-        return d
-
-    @threaded_validate(RegistrationSchema)
-    def put(self, rinfo, uaid=None, chid=None):
-        # type: (RouterInfo, Optional[uuid.UUID], Optional[str]) -> Deferred
-        """HTTP PUT
-
-        Update router type/data for a UAID.
+        uaid = uuid.uuid4()
+        chid = router_data.get("channelID", uuid.uuid4().hex)
```

Comment:
Please add warning that the value must match what's specified by the bridged client. We were bit pretty hard by this one and I'd like to make sure it doesn't come back.